Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,603 @@
+                preconditionFailure("If we receive a response body, we must have received a head before")
+
+            case .executing(let context, let requestState, .buffering(let streamID, var currentBuffer, next: let next)):
+                guard case .askExecutorForMore = next else {
+                    preconditionFailure("If we have received an error or eof before, why did we get another body part? Next: \(next)")
+                }
+
+                if currentBuffer.isEmpty {
+                    currentBuffer = buffer
+                } else {
+                    currentBuffer.append(contentsOf: buffer)
+                }
+                self.state = .executing(context, requestState, .buffering(streamID, currentBuffer, next: next))
+                return .none
+
+            case .executing(let executor, let requestState, .waitingForStream(var currentBuffer, next: let next)):
+                guard case .askExecutorForMore = next else {
+                    preconditionFailure("If we have received an error or eof before, why did we get another body part? Next: \(next)")
+                }
+
+                if currentBuffer.isEmpty {
+                    currentBuffer = buffer
+                } else {
+                    currentBuffer.append(contentsOf: buffer)
+                }
+                self.state = .executing(executor, requestState, .waitingForStream(currentBuffer, next: next))
+                return .none
+
+            case .executing(let executor, let requestState, .waitingForRemote(let streamID, let continuation)):
+                var buffer = buffer
+                let first = buffer.removeFirst()
+                self.state = .executing(executor, requestState, .buffering(streamID, buffer, next: .askExecutorForMore))
+                return .succeedContinuation(continuation, first)
+
+            case .finished:
+                // the request failed or was cancelled before, we can ignore further data
+                return .none
+
+            case .executing(_, _, .finished):
+                preconditionFailure("Received response end. Must not receive further body parts after that. Invalid state: \(self.state)")
+            }
+        }
+
+        enum ResponseBodyDeinitedAction {
+            case cancel(HTTPRequestExecutor)
+            case none
+        }
+
+        mutating func responseBodyDeinited() -> ResponseBodyDeinitedAction {
+            switch self.state {
+            case .initialized,
+                 .queued,
+                 .executing(_, _, .initialized):
+                preconditionFailure("Invalid state: \(self.state)")
+
+            case .executing(_, _, .waitingForStream(_, next: .endOfFile)):
+                self.state = .finished(error: nil, nil)
+                return .none
+
+            case .executing(let context, _, .waitingForStream(_, next: .askExecutorForMore)):
+                self.state = .finished(error: nil, nil)
```

Comment:
Why is this not an error?

Reply:
It doesn't actually matter I assume. Everything will be cleaned up after this anyways, and the user isn't interested in the response body. Happy to change to `HTTPClientError.cancelled`.